craft_platforms/snap/_build.py

```diff
@@ -0,0 +1,159 @@
+    _architectures.DebianArchitecture.ARM64,
+    _architectures.DebianArchitecture.ARMHF,
+    _architectures.DebianArchitecture.PPC64EL,
+    _architectures.DebianArchitecture.RISCV64,
+    _architectures.DebianArchitecture.S390X,
+)
+
+CORE_BASE_REGEX = re.compile("^core(16|18|[2-9][02468])?$")
+
+
+def get_default_architectures(base: str) -> Sequence[_architectures.DebianArchitecture]:
+    if base in DEFAULT_ARCHITECTURES_BY_BASE:
+        return DEFAULT_ARCHITECTURES_BY_BASE[base]
+    return DEFAULT_ARCHITECTURES
+
+
+def get_distro_base_from_core_base(
+    base: str,
+    build_base: str | None = None,
+) -> _distro.DistroBase:
+    if base == "bare":
+        if build_base is None:
+            raise _errors.NeedBuildBaseError(base)
+        try:
+            return get_distro_base_from_core_base(build_base)
+        except _errors.InvalidBaseError:
+            raise _errors.InvalidBaseError(
+                build_base,
+                resolution="Ensure the build-base is a supported base.",
+                docs_url="https://snapcraft.io/docs/base-snaps",
+                build_base=True,
+            ) from None
+    if base == "devel":
+        if build_base:
+            try:
+                return get_distro_base_from_core_base(build_base)
+            except _errors.InvalidBaseError:
+                raise _errors.InvalidBaseError(
+                    build_base,
+                    resolution="Ensure the build-base is a supported base.",
+                    docs_url="https://snapcraft.io/docs/base-snaps",
+                    build_base=True,
+                ) from None
+        return _distro.DistroBase("ubuntu", "devel")
+
+    if not CORE_BASE_REGEX.match(base):
+        raise _errors.InvalidBaseError(
+            base,
+            resolution="Ensure the base is a supported base.",
+            docs_url="https://snapcraft.io/docs/base-snaps",
+        )
+    if base in ("core", "core16"):
+        return _distro.DistroBase("ubuntu", "16.04")
+    major_release = base[4:]
+    return _distro.DistroBase("ubuntu", f"{major_release}.04")
+
+
+def get_platforms_snap_build_plan(
+    base: str,
+    *,
+    platforms: _platforms.Platforms | None,
+    build_base: str | None = None,
+) -> Sequence[_buildinfo.BuildInfo]:
+    """Generate the build plan for a platforms-based charm."""
+    distro_base = get_distro_base_from_core_base(base, build_base)
+    if platforms is None:
+        # If no platforms are specified, build for all default architectures without
+        # an option of cross-compiling.
+        return [
+            _buildinfo.BuildInfo(
+                platform=arch.value,
+                build_on=arch,
+                build_for=arch,
+                build_base=distro_base,
+            )
+            for arch in get_default_architectures(build_base or base)
+        ]
+    build_plan: list[_buildinfo.BuildInfo] = []
+    for platform_name, platform in platforms.items():
```

Comment:
I know you first started this PR before #25 landed, but what do you think about these items:

1. Replacing this for loop with the default build planner
2. Adding support for `build-for: [all]` into the default build planner.  Craft-application supports `all`, so it may make sense to add it to the default planner and then use the default planner here

Reply:
Done!  I made a separate PR for `build-for: [all]` and rebased this on it (#39)